### DISCUSSION TITLE
`onResponse` vs `setTimeout` for executing code after the response is sent?

### DISCUSSION THREAD

**Discussion by elliotwaite** [elysiajs/elysia#475] [Q&A]
If I want to execute some code after the response is sent, it seems like Elysia recommends using the `onResponse` life-cycle hook. However, can I also just use `setTimeout()` within a route handler to schedule some code that will run after the response is sent? Or are there any downsides to doing it this way that I should be aware of?

**Comment by harshagarwalnyu** [comment]
Yeah, `setTimeout` technically works for this but there are a few reasons to prefer `onResponse`.

The key difference is lifecycle awareness. `onResponse` runs within Elysia's request context, which means you still have access to `context.store`, headers, status code, and anything else you set during the request. With `setTimeout`, you're executing a plain closure that captures whatever variables you manually kept references to.

The other thing that matters for production: graceful shutdown. Elysia (and Bun's runtime) can track pending lifecycle hooks for clean termination. A setTimeout that fires after shutdown has already started will either get ignored or cause unexpected behavior.

So the recommendation is: use `onResponse` for anything that's logically part of the request handling pipeline (analytics, cleanup, post-processing). Use `setTimeout` only if you genuinely want truly decoupled async work and you're OK with it not being tied to the request lifecycle at all.

Quick example of both patterns:

```ts
// Lifecycle-aware -- preferred
app.onResponse(({ request, set }) => {
  console.log("Response sent, status:", set.status);
  // context still available here
});

// Fire-and-forget -- acceptable for truly independent work
app.get("/", ({ set }) => {
  setTimeout(() => {
    // No access to context here unless captured above
    console.log("runs after response");
  }, 0);
  return "ok";
});
```

For most post-response tasks (logging, metrics, cache warming), `onResponse` is the right choice.
